File: app/src/main/java/com/mudrichenko/evgeniy/flickrtestproject/ui/photoInfo/PhotoInfoPresenter.java
package com.mudrichenko.evgeniy.flickrtestproject.ui.photoInfo;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;
import com.mudrichenko.evgeniy.flickrtestproject.App;
import com.mudrichenko.evgeniy.flickrtestproject.R;
import com.mudrichenko.evgeniy.flickrtestproject.data.repository.PhotoInfoRepository;
import com.mudrichenko.evgeniy.flickrtestproject.database.FlickrPhoto;
import com.mudrichenko.evgeniy.flickrtestproject.utils.NetworkUtils;
import com.orhanobut.logger.Logger;

import org.jetbrains.annotations.NotNull;

import javax.inject.Inject;

@InjectViewState
public class PhotoInfoPresenter extends MvpPresenter<PhotoInfoView>
        implements PhotoInfoRepository.PhotoInfoRepositoryListener {

    @Inject
    PhotoInfoRepository mPhotoInfoRepository;

    @Inject
    NetworkUtils mNetworkUtils;

    public PhotoInfoPresenter() {
        App.Companion.getAppComponent().inject(this);
    }

    public void onPhotoInfoLoad(final long photoId) {
        if (!mNetworkUtils.isInternetConnectionAvailable()) {
            getViewState().showInfoMessage(App.Companion.getAppContext().getResources().getString(R.string.no_internet_connection));
            return;
        }
        getViewState().hideInfoMessage();
        getViewState().showProgressWheel();
        mPhotoInfoRepository.setPhotoInfoRepositoryListener(this);
        mPhotoInfoRepository.onPhotoInfoLoad(photoId);
    }

    public void onButtonDeleteClick(final long photoId) {
        mPhotoInfoRepository.setPhotoInfoRepositoryListener(this);
        mPhotoInfoRepository.onPhotoDeleteFromServer(photoId);
    }

    @Override
    public void onErrorReceived(String errorMessage) {
        getViewState().hideProgressWheel();
        getViewState().showInfoMessage(errorMessage);
    }

    @Override
    public void onPhotoDeleted() {
        Logger.i("PhotoInfoPresenter; onPhotoDeleted");
        getViewState().onPhotoDeleted();
    }

    public void unsubscribe() {
        mPhotoInfoRepository.unsubscribe();
    }

    @Override
    public void onPhotoInfoReceived(@NotNull FlickrPhoto flickrPhoto) {
        getViewState().showPhotoInfo(flickrPhoto);
        getViewState().hideProgressWheel();
        getViewState().hideInfoMessage();
    }

}

File: app/src/main/java/com/mudrichenko/evgeniy/flickrtestproject/ui/photoInfo/PhotoInfoPresenter.kt
package com.mudrichenko.evgeniy.flickrtestproject.ui.photoInfo

import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import com.mudrichenko.evgeniy.flickrtestproject.App
import com.mudrichenko.evgeniy.flickrtestproject.R
import com.mudrichenko.evgeniy.flickrtestproject.data.repository.PhotoInfoRepository
import com.mudrichenko.evgeniy.flickrtestproject.database.FlickrPhoto
import com.mudrichenko.evgeniy.flickrtestproject.utils.NetworkUtils
import com.orhanobut.logger.Logger

import javax.inject.Inject

@InjectViewState
class PhotoInfoPresenter : MvpPresenter<PhotoInfoView>(), PhotoInfoRepository.PhotoInfoRepositoryListener {

    @Inject
    lateinit var mPhotoInfoRepository: PhotoInfoRepository

    @Inject
    lateinit var mNetworkUtils: NetworkUtils

    init {
        App.appComponent!!.inject(this)
    }

    fun onPhotoInfoLoad(photoId: Long) {
        if (!mNetworkUtils.isInternetConnectionAvailable()) {
            viewState.showInfoMessage(App.appContext!!.resources.getString(R.string.no_internet_connection))
            return
        }
        viewState.hideInfoMessage()
        viewState.showProgressWheel()
        mPhotoInfoRepository.setPhotoInfoRepositoryListener(this)
        mPhotoInfoRepository.onPhotoInfoLoad(photoId)
    }

    fun onButtonDeleteClick(photoId: Long) {
        mPhotoInfoRepository.setPhotoInfoRepositoryListener(this)
        mPhotoInfoRepository.onPhotoDeleteFromServer(photoId)
    }

    override fun onErrorReceived(errorMessage: String) {
        viewState.hideProgressWheel()
        viewState.showInfoMessage(errorMessage)
    }

    override fun onPhotoDeleted() {
        Logger.i("PhotoInfoPresenter; onPhotoDeleted")
        viewState.onPhotoDeleted()
    }

    fun unsubscribe() {
        mPhotoInfoRepository.unsubscribe()
    }

    override fun onPhotoInfoReceived(flickrPhoto: FlickrPhoto) {
        viewState.showPhotoInfo(flickrPhoto)
        viewState.hideProgressWheel()
        viewState.hideInfoMessage()
    }

}
